<h2>Основные методы</h2>

<!-- xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx -->
<h3>1. Создание</h3>
<!-- xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx -->
<ul class="list-point">
	<li>Внути new Promise любые асинхронные операции</li>
	<li>Промис содержит свое состояние</li>
	<li><u>resolve</u> и <u>reject</u> функции</li>
    <li><u>resolve(результат)</u> при успешном выполнении <u>.then</u></li>
    <li><u>reject(ошибка)</u> при ошибке <u>.catch</u></li>
</ul>

<pre><code class="js border">
let promise = new Promise(function(resolve, reject) {
	resolve(arg) // => onFulfilled(arg) => .then(arg)
	reject(arg)  // => onRejected(arg)  => .catch(arg)
});
</code></pre>

<!------------------------------------------------------------->
<h4>Promise</h4>
<!------------------------------------------------------------->
<pre><code class="js"><u-code-title>Объявление</u-code-title>
let promise = new <u-code>Promise</u-code>((resolve, reject) => {
    if (...) {
    	// onResolved()
        <u-code>resolve</u-code>('ok');
    } else {
    	// onRejected()
        <u-code>reject</u-code>(new Error('Api error')); // аналогично с throw
        throw new Error('Api error');   // аналогично с reject
    }
});
</code></pre>

<pre><code class="js"><u-code-title>Использование</u-code-title>
promise
    <u-code>.then</u-code>(result => {
        console.log(result);
    })
    <u-code>.catch</u-code>(error => {
        console.error(error);
        console.error(error.message); 
    })
</code></pre>

<!------------------------------------------------------------->
<h4>Функция</h4>
<!------------------------------------------------------------->
<pre><code class="js"><u-code-title>Объявление</u-code-title>
function asyncFunc() {
	let promise = new Promise((resolve, reject) => {
		//
	});
	return promise;
}
</code></pre>

<pre><code class="js"><u-code-title>Использование</u-code-title>
asyncFunc()
    .then(result => { ... })
    .catch(result => { ... })
</code></pre>


<!-- xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx -->
<h3>2. Обработка Promise в асинхронных операциях</h3>
<!-- xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx -->

<!------------------------------------------------------------->
<h4>1. Один then</h4>
<!------------------------------------------------------------->
<pre><code class="js">
asyncFunc()
	<u-code>.then</u-code>(functionResolve, functionReject);
	// functionResolve - функция, которая сработает если обещание выполняется
	// functionReject  - функция, которая сработает если обещание не будет выполнено
</code></pre>

<pre><code class="js">
promise.then(
	resolved => console.log(resolved),
	error => console.log(error)
);
</code></pre>




<pre><code class="js"><u-code-title>js</u-code-title>
asyncFunc()
	.then((result) => { 
		console.log(result); 
	})
	.catch((error) => { 
		console.log(error);
	})

asyncFunc()
	.then((result) => { 
		console.log(result); 
	}, (error) => {
		console.log(error);
	})

</code></pre>

.then принимает 2параметра. 1й вызывается в случае успеха. 2й вызывается в случае ошибки


<!------------------------------------------------------------->
<h4>2. Цепочка then</h4>
<!------------------------------------------------------------->
<pre><code class="js">
asyncFunc()
	<u-code>.then</u-code>(functionResolve1)
	<u-code>.then</u-code>(functionResolve2)
	<u-code>.catch</u-code>(functionReject);
</code></pre>

<!------------------------------------------------------------->
<h4>3. Вызванная функция</h4>
<!------------------------------------------------------------->
<pre><code class="js">
asyncFunc()
	<u-code>.then</u-code>(functionResolve1()) // вызванная функция пропускается
</code></pre>
